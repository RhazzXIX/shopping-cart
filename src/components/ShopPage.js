import uniqid from "uniqid";
import { useState, useEffect } from "react";
import ProductCard from "./ProductCard";
import "../styles/ShopPage.css";
import cartImg from "../images/cart.svg";

const modifyCategoryNames = (string) => {
  return string
    .slice(0, 1)
    .toUpperCase()
    .concat(string.slice(1))
    .split("-")
    .join(" ");
};

const ShopPage = () => {
  const [categories, setCategories] = useState([]);
  const [showAllItems, setShowAllItems] = useState(true);
  const [products, setProducts] = useState([]);
  const [categoryNeeded, setCategoryNeeded] = useState("");
  const [cartItems, setCartItems] = useState([]);

  const getAllItems = (e) => {
    setShowAllItems(true);
  };

  const getCategories = async () => {
    const fetchedCategories = [];
    const data = await fetch("https://dummyjson.com/products/categories").then(
      (res) => res.json()
    );
    data.forEach((category) => {
      const fetchedCategory = {
        name: category,
        id: uniqid(),
      };
      fetchedCategories.push(fetchedCategory);
    });
    setCategories(fetchedCategories);
  };

  const setCategory = (e) => {
    const [neededCategory] = categories.filter(
      (category) => category.id === e.target.dataset.id
    );
    setCategoryNeeded(neededCategory.name);
    setShowAllItems(false);
  };

  const getProducts = async () => {
    let url = `https://dummyjson.com/products/category/${categoryNeeded}`;
    if (showAllItems) url = "https://dummyjson.com/products?limit=0";
    const data = await fetch(url).then((res) => res.json());
    setProducts(data.products);
  };

  useEffect(() => {
    getCategories();
    getProducts();
  }, []);

  useEffect(() => {
    getProducts();
  }, [categoryNeeded, showAllItems]);

  return (
    <main id="shop">
      <section id="sidebar">
        <h3>Categories</h3>
        <ul>
          <li>
            <button onClick={getAllItems}>All Items</button>
          </li>
          {categories.map((category) => {
            return (
              <li key={category.id}>
                <button onClick={setCategory} data-id={category.id}>
                  {modifyCategoryNames(category.name)}
                </button>
              </li>
            );
          })}
        </ul>
        <div id="e-com">
          <p>E-commerce data is powered by:</p>
          <a href="https://dummyjson.com/">DummyJSON </a>
        </div>
      </section>
      <section id="items">
        <header>
          <h2>
            {showAllItems ? "All Items" : modifyCategoryNames(categoryNeeded)}
          </h2>
          <button id="cart">
            <img src={cartImg} alt="Cart" />
            {cartItems.length !== 0 && <p>{cartItems.length}</p>}
          </button>
        </header>
        <div className="productContainer">
          {products.map((product) => (
            <ProductCard product={product} key={product.id} />
          ))}
        </div>
      </section>
    </main>
  );
};

export default ShopPage;
