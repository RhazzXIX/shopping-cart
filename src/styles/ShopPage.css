main#shop {
  display: flex;
  gap: 4px;
  flex: auto;
}

section#sidebar {
  padding: 10px 20px;
  box-shadow: 2px 0 8px black;
  background-color: rgb(230, 230, 230);
  min-width: 210px;
  display: grid;
  grid-template-rows: 40px auto auto;
}

section#sidebar > ul {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 8px;
}

section#sidebar > ul button {
  appearance: none;
  border: 0;
  font-size: 1.1rem;
  letter-spacing: 1px;
  cursor: pointer;
  background-color: rgb(230, 230, 230);
}

section#sidebar > ul button:hover {
  text-decoration: underline;
}

section#sidebar > ul button:active {
  color: grey;
}

div#e-com {
  align-self: end;
  width: 148px;
  border-radius: 8px;
  font-size: 0.8rem;
  background-color: rgba(255, 255, 255, 0.6);
  padding: 4px 8px;
}

div#e-com:hover {
  background-color: white;
}

section#items {
  padding: 8px 16px;
  display: flex;
  flex-direction: column;
  gap: 16px;
  flex: auto;
  max-height: 821px;
}

section#items > header {
  border-bottom: 2px solid black;
  box-shadow: 0 8px 8px -8px black;
  padding: 6px;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

button#cart {
  width: 70px;
  height: 40px;
  border: 1px solid black;
  border-radius: 8px;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  background-color: white;
}

button#cart:hover {
  box-shadow: 0 0 4px black;
}

button#cart:active {
  filter: invert(100);
}

button#cart img {
  height: 30px;
  width: 30px;
}

div.productContainer {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  gap: 8px;
  padding: 4px;
  overflow: auto;
  flex: auto;
}
