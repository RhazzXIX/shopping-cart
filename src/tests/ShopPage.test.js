import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import ShopPage from "../components/ShopPage";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

const user = userEvent.setup();

describe("ShopPage component", () => {
  it("renders the component", () => {
    const { container } = render(<ShopPage />);
    expect(container).toMatchSnapshot();
  });

  describe("Category section", () => {
    it("renders fetched category", async () => {
      render(<ShopPage />);
      expect(screen.queryByRole("button", { name: "Smartphones" })).toBeNull();
      const button = await screen.findByRole("button", { name: "Smartphones" });
      expect(button).toBeInTheDocument();
    });
  });

  describe("Product Section", () => {
    it("renders the fetched products", async () => {
      render(<ShopPage />);
      const productDetailButtons = await screen.findAllByRole("button", {
        name: "View Details",
      });
      expect(productDetailButtons).toHaveLength(100);
    });

    it("renders the right product section when the user click in a category", async () => {
      render(<ShopPage />);
      const productSection = screen.getByRole("heading", {
        name: "All Items",
      });
      expect(productSection.textContent).toMatch("All Items");
      const button = await screen.findByRole("button", { name: "Smartphones" });
      expect(button).toBeInTheDocument();
      await user.click(button);
      expect(productSection.textContent).toMatch("Smartphones");
    });
    it(`let's the user shuffle between categories`, async () => {
      render(<ShopPage />);
      const allItemsBtn = screen.getByRole("button", { name: "All Items" });
      const smartphoneBTN = await screen.findByRole("button", {
        name: "Smartphones",
      });
      const womenShoesBtn = await screen.findByRole("button", {
        name: "Womens shoes",
      });
      let treeOIl = await screen.findByRole("heading", {
        name: "Tree Oil 30ml",
        level: 4,
      });
      let keyHolder = await screen.findByRole("heading", {
        name: "Key Holder",
        level: 4,
      });
      let sneakerShoes = await screen.findByRole("heading", {
        name: "Sneaker shoes",
        level: 4,
      });
      let productDetailButtons = await screen.findAllByRole("button", {
        name: "View Details",
      });
      expect(productDetailButtons).toHaveLength(100);
      expect(treeOIl).toBeInTheDocument();
      expect(keyHolder).toBeInTheDocument();
      expect(sneakerShoes).toBeInTheDocument();
      await user.click(smartphoneBTN);
      await waitForElementToBeRemoved(
        screen.queryByRole("heading", { name: "Tree Oil 30ml", level: 4 })
      ).then(async () => {
        productDetailButtons = await screen.findAllByRole("button", {
          name: "View Details",
        });
        expect(treeOIl).not.toBeInTheDocument();
        expect(keyHolder).not.toBeInTheDocument();
        expect(sneakerShoes).not.toBeInTheDocument();
        expect(productDetailButtons).toHaveLength(5);
      });
      await user.click(womenShoesBtn);
      sneakerShoes = await screen.findByRole("heading", {
        name: "Sneaker shoes",
        level: 4,
      });
      expect(sneakerShoes).toBeInTheDocument();
      expect(treeOIl).not.toBeInTheDocument();
      expect(keyHolder).not.toBeInTheDocument();
      await user.click(allItemsBtn);
      keyHolder = await screen.findByRole("heading", {
        name: "Key Holder",
        level: 4,
      });
      treeOIl = await screen.findByRole("heading", {
        name: "Tree Oil 30ml",
        level: 4,
      });
      productDetailButtons = await screen.findAllByRole("button", {
        name: "View Details",
      });
      expect(keyHolder).toBeInTheDocument();
      expect(treeOIl).toBeInTheDocument();
      expect(sneakerShoes).toBeInTheDocument();
      expect(productDetailButtons).toHaveLength(100);
    });
  });
});
