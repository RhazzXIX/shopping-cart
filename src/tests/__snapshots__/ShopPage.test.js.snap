// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage component renders the component 1`] = `
<div>
  <main
    id="shop"
  >
    <section
      id="sidebar"
    >
      <h3>
        Categories
      </h3>
      <ul>
        <li>
          <button>
            All Items
          </button>
        </li>
      </ul>
    </section>
    <section
      id="items"
    >
      <header>
        <h2>
          All Items
        </h2>
        <button
          id="cart"
        >
          <img
            alt="Cart"
            src="cart.svg"
          />
        </button>
      </header>
      <div
        class="productContainer"
      />
    </section>
  </main>
</div>
`;
